src/gravity/gravity.cpp

```diff
@@ -20,6 +20,30 @@
 using namespace parthenon::package::prelude;
 
 namespace Gravity {
+
+struct GravParams {
+  Real gm, tstart, tstop;
+  GravParams(ParameterInput *pin) {
+    gm = pin->GetOrAddReal("gravity", "gm", 0.0);
+    tstart = pin->GetOrAddReal("gravity", "tstart", std::numeric_limits<Real>::lowest());
+    tstop = pin->GetOrAddReal("gravity", "tstop", std::numeric_limits<Real>::max());
+  }
+  void overwrite(std::string block_name, ParameterInput *pin) {
+    if (pin->DoesParameterExist(block_name, "gm")) {
+      gm = pin->GetReal(block_name, "gm");
+      pin->SetReal("gravity", "gm", gm);
```

Comment:
As you know from our history with user-defined BCs, I'd really prefer to not manually populate `pin` in this manner.  Can we think of an alternative design?  I was actually not opposed to the `type` specification for gravity before, but your PR seems to imply that the documentation itself is mandating the code design here.  I am sure we could equally find an alternative solution with Sphinx.

Reply:
The user-defined BCs were different because that was modifying "parthenon" parameters. These are "our" parameters that we can do whatever we want to (IMO). 

I'm doing this for three reasons: 
  (1) there are common parameters between the types. Those parameters can go in a ` `<gravity>` block and be retrieved by other packages from the `<gravity>` block without figuring out which gravity type is active.

  (2) because other packages at `Initialize` time require `GM`, and the easiest way to get `GM` is to read it from `<gravity>`. I understand that we can pass `GM` through to the other `Initialize` functions, but I don't like the idea of `Gas::Initialize` needing `GM`, and then passing that to the viscosity initializer. 

  (3) If I don't do this, then either we have repeated code for the common parameters, or we force them under `<gravity>`. In that case, you will run into stuff like this:
```
<gravity>
gm = 1.0
<gravity/point>
```
with an "empty" node. I'd prefer this instead:
```
<gravity/point>
gm = 1.0
```
and then we put the `gm` value into the `<gravity>` so that everything is consistent and other packages can freely read `gm` without having to know the type.  

I think this provides a much cleaner interface to the user.

Reply:
I guess my opinion would be that the existing `type` design is a cleaner interface to the developer and user --- _except_ for documentation... but is there not some way in Sphinx to have an entry (e.g., `type = uniform`) that reveals a unique set of additional parameters underneath? We already have a special key `opt` to distinguish these entries.  I think it might be interesting to pursue this opportunity rather than overhauling the gravity package as here...

> These are "our" parameters that we can do whatever we want to (IMO).

Equally, perhaps, it is "my" input file.  And I don't feel comfortable with `artemis` manipulating `ParameterInput` behind the scenes.  Also, I feel like one has to have a very solid understanding of `artemis` and `parthenon` to understand how this kind of design choice melds with other code ops, e.g., the expected vs desired behavior for these:

`./artemis -i my_input.in`
`./artemis -i my_input.in gravity/gm=2.0`
`./artemis -i my_input.in gravity/point/gm=2.0`
`./artemis -i my_input.in gravity/gm=2.0 gravity/point/gm=123456789`
`./artemis -r my_restart.rhdf`
`./artemis -r my_restart.rhdf gravity/gm=2.0`
`./artemis -r my_restart.rhdf gravity/point/gm=2.0`
`./artemis -r my_restart.rhdf gravity/gm=2.0 gravity/point/gm=123456789`
`./artemis -r my_restart.rhdf -i my_input.in`
`./artemis -r my_restart.rhdf -i my_input.in gravity/gm=2.0`
`./artemis -r my_restart.rhdf -i my_input.in gravity/point/gm=2.0`
`./artemis -r my_restart.rhdf -i my_input.in gravity/gm=2.0 gravity/point/gm=123456789`

I don't think I immediately know the answer to all of them... And even with appropriate guards against some of these, I still think there are questions surrounding what each of these _should_ give.

Finally, I'll just note that we have already taken steps to eliminate this pattern before... I'd suggest we don't re-introduce it here just for documentation's sake.... but.... of course, this is a team effort: rather than just accepting my proposal or just going with your's, is there some intermediate solution that makes everybody happy? That would be ideal 😸

Reply:
I would say the main reason we manipulated the input parameter file in the past was to work around a deficiency in Parthenon itself. And it was important that we were able to do that, but modifying the input parameters sight-unseen can in my opinion reasonably be expected to sometimes surprise users, and I think that's something we should be aggressive about avoiding. It can also lead to some weirdness with processing parameters during restarts, and I actually don't know what it does in @pdmullen's example above with runtime input.

My feeling is we should avoid that pattern when possible. I actually like the empty node approach:
```
<gravity>
gm = 1.0 
<gravity/point>
```
I think it shows we are serious about the node-based nature of input parameters. It also avoids any weirdness about multiple inconsistent `gm` values, and makes clear that `gm` means the exact same thing for every gravity sub-node that uses that parameter.

It would actually be nice if package `Initialize` functions could call packages already initialized -- I think that would ease some of the tension that drives modifying the input parameter file (it's done in `phoebus` in a few places for example) -- but I don't know that that would be easy to set up.

Reply:
`Initialize` can take as argument the existing `packages` object (I believe KHARMA does this IIRC?)... but this does impose a certain ordering to initialize packages, if, for example, we read something from the `gravity` package params in the `gas` package.  Only tangentially related to this thread, but thought I'd mention it anyways.

Reply:
Deal 🤝

Reply:
I agree about not using `type`, and this compromise looks good to me.

Reply:
Done